### PR TITLE
Feature: Sampler

### PR DESCRIPTION
**Pull request by JakobEliasWagner**
# Feature: Sampler

## Description

This pull request introduces base classes for sampling and an example (namely a uniform box-sampler). Samplers should cater to both continuous and discrete domains. By incorporating abstract base classes `Sampler` and `BoxSampler`, alongside a concrete implementation `UniformBoxSampler`, this pr introduces the groundwork for a flexible and extensible sampling module.
Sampling for operator learning is not only relevant for creating discrete representations of functions but could also be used to extend the capabilities of operators (resampling inside the operator). Therefore should this feature not be tied to the `data` module.

### Which issue does this PR tackle?

  -  Lack of a unified structure for sampling strategies.

### How does it solve the problem?

  - Creation of a new sub-module `discrete`.
  - Abstract Base Classes: The introduction of `Sampler` and `BoxSampler` abstract base classes. These classes outline essential methods that all derived samplers must implement, such as sampling methods and initialization routines.
  - `Sampler` and `BoxSampler` are distinct classes as not all domains are boxes (spheres, tori, or more complex geometries). 
  - `UniformBoxSampler`: Serves as a concrete implementation of the `BoxSampler` abstract class, offering a ready-to-use sampler that uniformly samples within specified bounds. This addition demonstrates the practical application of the new abstract classes.

### How are the changes tested?

  - Unittests:
    - intialization
    - delta
    - within bounds sampling
    - is uniform distribution

## Checklist for Contributors

- [x] Scope: This PR tackles exactly one problem.
- [x] Conventions: The branch follows the `feature/title-slug` convention.
- [x] Conventions: The PR title follows the `Bugfix: Title` convention.
- [x] Coding style: The code passes all pre-commit hooks.
- [x] Documentation: All changes are well-documented.
- [x] Tests: New features are tested and all tests pass successfully.
- [x] Changelog: Updated CHANGELOG.md for new features or breaking changes.
- [x] Review: A suitable reviewer has been assigned.

## Checklist for Reviewers:

- [x] The PR solves the issue it claims to solve and only this one.
- [x] Changes are tested sufficiently and all tests pass.
- [x] Documentation is complete and well-written.
- [x] Changelog has been updated, if necessary.

### CODE REVIEW

**Comment by JakobEliasWagner** [comment]
Freeze this branch for now, until features have been added onto other branches.

File: src/continuity/discrete/box_sampler.py [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+
+from abc import ABC
+
+import torch
+
+from .sampler import Sampler
+
+
+class BoxSampler(Sampler, ABC):
+    r"""Abstract base class for sampling from n-dimensional boxes.
```
Comment (samuelburbulla) [inline src/continuity/discrete/box_sampler.py:15]: Simplified this docstring.

File: src/continuity/discrete/sampler.py [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+
+    Args:
+        ndim: Dimension of the domain.
+    """
+
+    def __init__(self, ndim: int):
+        self.ndim = ndim
+
+    @abstractmethod
+    def __call__(self, n: int) -> torch.Tensor:
```
Comment (samuelburbulla) [inline src/continuity/discrete/sampler.py:26]: Renamed `n_samples` to `n`.

File: src/continuity/discrete/uniform.py [hunk 3]
```diff
@@ -0,0 +1,41 @@
+"""
+`continuity.discrete.uniform`
+
+Uniform samplers.
```
Comment (samuelburbulla) [inline src/continuity/discrete/uniform.py:4]: I guess this module will also contain other uniform samplers, not necessarily box samplers only.

File: src/continuity/discrete/uniform.py [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+        sampler = UniformBoxSampler(torch.tensor([0, 0]), torch.tensor([1, 1]))
+        samples = sampler(100)
+        samples.shape
+        ```
+        Output:
+        ```
+        torch.Size([100, 2])
+        ```
+
+    Note:
```
Comment (samuelburbulla) [inline src/continuity/discrete/uniform.py:26]: Simplified the docstring (which was contradicting itself) and added this note instead.

File: tests/discrete/test_uniform.py [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+    for sampler in sampler_list:
+        samples = sampler(n_samples)
+        samples_min, _ = samples.min(dim=0)
+        samples_max, _ = samples.max(dim=0)
+        assert torch.greater_equal(samples_min, sampler.x_min).all()
+        assert torch.less_equal(samples_max, sampler.x_max).all()
+
+
+def test_uniform_distribution(sampler_list):
+    n_samples = 2**20
```
Comment (samuelburbulla) [inline tests/discrete/test_uniform.py:45]: Increased number of samples as test failed locally from time to time.

**Review by samuelburbulla — APPROVED** [review]
Pushed some comments directly. Feel free to merge if you agree.

**Comment by samuelburbulla** [comment]
We could also add a CHANGELOG entry.
